File: toplas2021/benchmarks/seahorn/5-linger-dec/linger-dec-11-1.c
#include "seahorn/seahorn.h"

extern int nd();

void linger_dec_bound_three(int n, int *pa, int *pb, int *pc) {
  if (n == 0) return;
  *pa -= 1;
  *pb -= 2;
  *pc -= 3;
  int x = nd();
  if (nd()) {
    pa = &x;
  } else if (nd()) {
    pb = &x;
  } else if (nd()) {
    pc = &x;
  }
  linger_dec_bound_three(n - 1, pa, pb, pc);
}

int main() {
  int n = nd(), a = nd(), b = nd(), c = nd();
  int old_a = a;
  linger_dec_bound_three(n, &a, &b, &c);
  sassert(old_a >= a && old_a - a < 3 * n);
}

File: esop2020/benchmarks/rust-horn/5-linger-dec/linger-dec-11-1.rs
pub fn rand<T>() -> T { unimplemented!() }

fn linger_dec_bound_three(n: i32, ma: &mut i32, mb: &mut i32, mc: &mut i32) {
  if n == 0 {
    return;
  }
  *ma -= 1;
  *mb -= 2;
  *mc -= 3;
  let mut x = rand();
  let mut new_ma = ma;
  let mut new_mb = mb;
  let mut new_mc = mc;
  if rand() {
    new_ma = &mut x;
  } else if rand() {
    new_mb = &mut x;
  } else if rand() {
    new_mc = &mut x;
  }
  linger_dec_bound_three(n - 1, new_ma, new_mb, new_mc);
}
fn main() {
  let n = rand();
  let mut a = rand();
  let mut b = rand();
  let mut c = rand();
  let old_a = a;
  linger_dec_bound_three(n, &mut a, &mut b, &mut c);
  assert!(old_a >= a && old_a - a < 3 * n);
}
